### PR TITLE
Add gelu kernel example

### PR DESCRIPTION
**Pull request by 36330**
新增examples gelu算子，基线对比加速比为 0.5x，但是目前v核已经满载，以下为精度与性能对比测试报告

# 算子评测报告

**评测代号**: cann_final_eval_20260818_173034
**评测时间**: 2026-08-18T17:30:34.600750
**设备**: npu:0
**框架版本**: V0.4.0
**评测集版本**: tasks-v0.4.0

## 概览

| 指标 | 数值 |
|------|------|
| 评测算子数 | 1 |
| 总用例数 | 20 |
| 通过用例数 | 20 |
| 失败用例数 | 0 |
| 通过率 | 100.00% |
| 综合得分 | 63.95 |

### Gelu（level1/gelu）

| 指标 | 数值 |
|------|------|
| 用例数 | 20 |
| 通过数 | 20 |
| 失败数 | 0 |
| 通过率 | 100.00% |
| 平均加速比 | 0.50x |
| 编译/运行得分 | 20.00 |
| 精度得分 | 30.00 |
| 性能得分 | 13.95 |
| 得分 | 63.95 |

| 用例ID | 状态 | 耗时(μs) | 加速比 | 精度误差 |
|--------|------|----------|--------|----------|
| level1/gelu_1 | ✅ | 8.62 | 0.79x | 0.000977 |
| level1/gelu_2 | ✅ | 79.62 | 0.22x | 0.000000 |
| level1/gelu_3 | ✅ | 315.66 | 0.20x | 0.000000 |
| level1/gelu_4 | ✅ | 395.58 | 0.70x | 0.000000 |
| level1/gelu_5 | ✅ | 422.20 | 0.84x | 0.000000 |
| level1/gelu_6 | ✅ | 12.50 | 0.57x | 0.000000 |
| level1/gelu_7 | ✅ | 8.58 | 0.83x | 0.001953 |
| level1/gelu_8 | ✅ | 12.88 | 0.54x | 0.000000 |
| level1/gelu_9 | ✅ | 923.90 | 0.19x | 0.000015 |
| level1/gelu_10 | ✅ | 12.52 | 0.55x | 0.000000 |
| level1/gelu_11 | ✅ | 30.96 | 0.23x | 0.000000 |
| level1/gelu_12 | ✅ | 12.66 | 0.54x | 0.000000 |
| level1/gelu_13 | ✅ | 178.80 | 0.24x | 0.000000 |
| level1/gelu_14 | ✅ | 23.72 | 0.58x | 0.000000 |
| level1/gelu_15 | ✅ | 30.40 | 0.21x | 0.000000 |
| level1/gelu_16 | ✅ | 48.30 | 0.22x | 0.000000 |
| level1/gelu_17 | ✅ | 16.64 | 0.64x | 0.000000 |
| level1/gelu_18 | ✅ | 40.94 | 0.25x | 0.000000 |
| level1/gelu_19 | ✅ | 16.68 | 0.65x | 0.000000 |
| level1/gelu_20 | ✅ | 177.70 | 1.10x | 0.000000 |

image

### CODE REVIEW

File: examples/cann-bench/gelu/gelu.py [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+
+
+def gen_input(shape, dtype_str, value_range):
+    torch_dtype = TORCH_DTYPE_MAP[dtype_str]
+    low, high = value_range
+    x = torch.empty(shape, dtype=torch.float32).uniform_(low, high).to(torch_dtype)
+    return x
+
+
+def run_gelu_kernel(x, approximate="none", block_N=BLOCK_N):
```
Comment (ShareableXue) [inline examples/cann-bench/gelu/gelu.py]: `approximate` is currently ignored. The public API defaults to "none", but all dtype branches dispatch to a tanh kernel, so gelu(x) does not implement the documented exact GELU semantics. 

The official cann-bench cases contain both modes. 

Please restore separate exact/tanh dispatch and add precision coverage for both modes.

File: examples/cann-bench/gelu/gelu.py [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+ATOL_MAP = {"float16": 1e-3, "bfloat16": 1e-2, "float32": 1e-5}
+
+
+def run_gelu(case_id, shape, dtype_str, approximate, value_range):
+    torch_dtype = TORCH_DTYPE_MAP[dtype_str]
+    lo, hi = value_range
+    x = torch.empty(shape, dtype=torch.float32).uniform_(lo, hi).to(torch_dtype).npu()
+
+    y = gelu(x, approximate=approximate)
+    ref = torch.nn.functional.gelu(x, approximate="tanh")
```
Comment (ShareableXue) [inline examples/cann-bench/gelu/gelu.py]: The test harness hardcodes "tanh" in both the golden and every test invocation, so it cannot detect the missing "none" implementation. 

Please cover the official exact/tanh matrix separately. For exact mode, compute the golden on CPU.

File: examples/cann-bench/gelu/gelu.py [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+                    next_off = base + (tile + 1) * block_N
+
+                    T.wait_flag("mte3", "mte2", nxt)
+                    T.copy(X[next_off], x_h[nxt, :])
+                    T.set_flag("mte2", "v", nxt)
+
+                    T.wait_flag("mte2", "v", cur)
+                    T.tile.cast(x_ub[cur, :], x_h[cur, :], _CAST_LOW2HIGH, tile_elem)
+                    T.tile.mul(z_ub, x_ub[cur, :], RSQRT_2)
+                    T.tile.erf(t0, z_ub)
```
Comment (ShareableXue) [inline examples/cann-bench/gelu/gelu.py]: The T.tile.erf interface does not exist for now, you can check it in this file:
https://github.com/tile-ai/tilelang-ascend/blob/5149acdb40b086d7cd84e30821f5759fa8699a14/tilelang/language/ascend_tile.py

File: examples/cann-bench/gelu/gelu.py [hunk 4]
```diff
@@ -0,0 +1,620 @@
...
+            y_ub = T.alloc_ub((2, tile_elem), cal_dtype)
+            x_h = T.alloc_ub((2, tile_elem), in_dtype)
+            y_h = T.alloc_ub((2, tile_elem), in_dtype)
+            t0 = T.alloc_ub((tile_elem,), cal_dtype)
+            t1 = T.alloc_ub((tile_elem,), cal_dtype)
+            t2 = T.alloc_ub((tile_elem,), cal_dtype)
+            z_ub = T.alloc_ub((tile_elem,), cal_dtype)
+            h0 = T.alloc_ub((tile_elem,), cal_dtype)
+            h1 = T.alloc_ub((tile_elem,), cal_dtype)
+            ones = T.alloc_ub((tile_elem,), cal_dtype)
```
Comment (ShareableXue) [inline examples/cann-bench/gelu/gelu.py:325]: The capacity of UB on A2/A3 is 192KB. 
With block_N=8192, the static demand for UB capacity may overflow. 

Please careful planning of UB usage is necessary; otherwise, a segmentation fault will occur during the JIT stage.

File: examples/cann-bench/gelu/gelu.py [hunk 5]
```diff
@@ -0,0 +1,620 @@
...
+    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: False,
+    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
+}
+
+
+@tilelang.jit(out_idx=[-1], pass_configs=pipe_configs)
+def _gelu_tanh_kernel(N, block_N, dtype="float32"):
+    tile_elem = block_N // VEC_NUM
+    total_tiles = (N + block_N - 1) // block_N
+    tiles_per_block = max(2, min(8, (total_tiles + 23) // 24))
```
Comment (ShareableXue) [inline examples/cann-bench/gelu/gelu.py]: 这里的设计意图应该是：
VEC_NUM=2 让同一个 AI Core 内的两个 AIV 并行处理一个 block_N=8192 tile 的两半，每个 AIV 负责 4096 个元素；同时强制 tiles_per_block >= 2，做double buffer，使每个 AIV 可以通过两套 UB 在相邻两个 tile 之间做ping-pong，流水线并行重叠 MTE2 搬入、Vector 计算和 MTE3 搬出。

这个设计在logical tile足够多时是合理的。例如有至少 48 个 tile 时，可以启动 24 个AI Core，每个cid的两个 AIV分别处理 tile 的两半，同时每个 cid 内还有两个 tile 可以形成双缓冲。

---

问题是当前无条件执行：
tiles_per_block = max(2, ...)
当 tile 数量不足以同时填满所有 AI Core 时，它会为了构造双缓冲而减少 cid 数量，牺牲核间并行度。

例如 N=196608 时一共有 24 个 8192-element tile，但当前计算得到：
tiles_per_block = 2
num_blocks = 12
因此只启动 12 个 cid，即使用 12 个AICore、共 24 个 AIV；另外 12 个AIcore、共 24 个 AIV 空闲。
如果使用 tiles_per_block=1, num_blocks=24，则可以让 24 个 cid、48 个 AIV 同时处理这 24 个 tile。

此外，这种策略还会对更小的输入还会产生无效 Vector 计算。例如 N=1024 时只有一个不完整 tile，但仍然安排两个 stage。两个 AIV 各对两个 4096-element buffer 执行完整的V核计算（mul/axpy/sigmoid 链），总调度 16384 个 elements，但是其中只有 1024 个结果有效。有效计算只有1/16
GM copy 虽然有 tail extent，不会越界，但无效 lanes 的 Vector 计算仍然会执行。

Comment (ShareableXue) [inline examples/cann-bench/gelu/gelu.py]: 这里建议根据总 tile 数区分调度策略：
tile 数不足时优先使用 tiles_per_block=1 占满更多 AI Core；
在已有足够 cid 填满所有 AIV 后，再让每个 cid 处理多个 tile 启用双缓冲。

对于远小于 8192 的输入，建议考虑自适应减小 block_N 或者 增加新的小tensor的kernel路径

File: examples/cann-bench/gelu/gelu.py [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+__all__ = ["gelu"]
+
+
+if __name__ == "__main__":
+    import torch.nn.functional as F
+
+    torch.manual_seed(0)
+    all_ok = True
+
+    print("=== Exact mode (approximate='none') ===")
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:481]: gelu有两种实现: tanh 和 erf, 建议分成两个用例文件, 代码更清晰

File: examples/cann-bench/gelu/gelu.py [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+
+    if orig_dtype == torch.float16:
+        if approximate == "tanh":
+            cache_key = (N, block_N, "fp16_cast_tanh")
+            if cache_key not in _KERNEL_CACHE:
+                _KERNEL_CACHE[cache_key] = _gelu_tanh_cast_kernel(N, block_N, in_dtype="float16", cal_dtype="float32")
+        else:
+            cache_key = (N, block_N, "fp16_tanh")
+            if cache_key not in _KERNEL_CACHE:
+                _KERNEL_CACHE[cache_key] = _gelu_tanh_kernel(N, block_N, dtype="float16")
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:442]: 应该调用_gelu_exact_as_cast_kernel

File: examples/cann-bench/gelu/gelu.py [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+    if not x.is_npu:
+        x = x.npu()
+    orig_dtype = x.dtype
+    N = x.numel()
+    x_flat = x.contiguous().view(-1)
+
+    if block_N is None:
+        block_N = _compute_block_n(N)
+
+    if orig_dtype == torch.float16:
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:434]: 下面的 if 可以使用 map 替换:
```python
_GELU_KERNEL_TABLE = {
    # (dtype, approximate)   →  (kernel 工厂, 关键参数)
    ("float32", "tanh"):    (_gelu_tanh_kernel,        dict(dtype="float32")),
    ("float32", "none"):    (_gelu_exact_as_kernel,    dict(dtype="float32")),
    ("float16", "tanh"):    (_gelu_tanh_cast_kernel,   dict(in_dtype="float16", cal_dtype="float32")),
    ("float16", "none"):    (_gelu_exact_as_cast_kernel, dict(in_dtype="float16", cal_dtype="float32")),
    ("bfloat16", "tanh"):   (_gelu_tanh_cast_kernel,   dict(in_dtype="bfloat16", cal_dtype="float32")),
    ("bfloat16", "none"):   (_gelu_exact_as_cast_kernel, dict(in_dtype="bfloat16", cal_dtype="float32")),
}
```

File: examples/cann-bench/gelu/gelu.py [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+def _compute_block_n(N):
+    if N <= 1024:
+        return 1024
+    elif N <= 8192:
+        return max(256, ((N + 255) // 256) * 256)
+    else:
+        return 8192
+
+
+def _compute_schedule(N, block_N, ub_per_tile):
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:41]: ub_per_tile × tiles_per_block > _UB_BUDGET 假设 tiles_per_block 个 tile 同时驻留——实际分配是固定 2 级双缓冲（与 tile 数无关）。后果：fp32-exact 在 N>8192 时 tiles_per_block 被压到 1，块内完全失去流水（256 行的死代码逻辑白写）；且模型本身还低估实际占用（9 buffer vs 实际 11 buffer：x_ub/y_ub 是 2 级——block_N=8192 时实际 176KB vs 模型 147KB，擦着 192KB 预算）。_gelu_exact_as_cast_kernel 里又有一套独立的 block_N 减半逻辑，两处不一致。

File: examples/cann-bench/gelu/gelu.py [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+            base = cid * tiles_per_block * block_N + vid * tile_elem
+
+            with T.Scope("V"):
+                T.set_flag("mte3", "mte2", 0)
+                T.set_flag("mte3", "mte2", 1)
+                T.wait_flag("mte3", "mte2", 0)
+                T.copy(X[base], x_ub[0, :])
+                T.set_flag("mte2", "v", 0)
+
+                for tile in T.unroll(0, tiles_per_block - 1):
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:75]: 与下面last_stage的公式一致, 可合并

File: examples/cann-bench/gelu/gelu.py [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+                    T.tile.sigmoid(y_ub[cur, :], t0)
+                    T.tile.mul(y_ub[cur, :], x_ub[cur, :], y_ub[cur, :])
+                    T.tile.cast(y_h[cur, :], y_ub[cur, :], _CAST_HIGH2LOW, tile_elem)
+                    T.set_flag("v", "mte3", cur)
+
+                    T.wait_flag("v", "mte3", cur)
+                    T.copy(y_h[cur, :], Y[cur_off])
+                    T.set_flag("mte3", "mte2", cur)
+
+                last_stage = (tiles_per_block - 1) % 2
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:167]: 与上面的公式一致, 可合并

File: examples/cann-bench/gelu/gelu.py [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+                    T.tile.add(y_ub[cur, :], y_ub[cur, :], 1.0)
+                    T.tile.mul(y_ub[cur, :], y_ub[cur, :], 0.5)
+                    T.tile.mul(y_ub[cur, :], x_ub[cur, :], y_ub[cur, :])
+                    T.set_flag("v", "mte3", cur)
+
+                    T.wait_flag("v", "mte3", cur)
+                    T.copy(y_ub[cur, :], Y[cur_off])
+                    T.set_flag("mte3", "mte2", cur)
+
+                last_stage = (tiles_per_block - 1) % 2
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:261]: 与下面last_stage的公式一致, 可合并

File: examples/cann-bench/gelu/gelu.py [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+                    T.tile.mul(y_ub[cur, :], y_ub[cur, :], 0.5)
+                    T.tile.mul(y_ub[cur, :], x_ub[cur, :], y_ub[cur, :])
+                    T.tile.cast(y_h[cur, :], y_ub[cur, :], _CAST_HIGH2LOW, tile_elem)
+                    T.set_flag("v", "mte3", cur)
+
+                    T.wait_flag("v", "mte3", cur)
+                    T.copy(y_h[cur, :], Y[cur_off])
+                    T.set_flag("mte3", "mte2", cur)
+
+                last_stage = (tiles_per_block - 1) % 2
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:381]: 与下面last_stage的公式一致, 可合并

File: examples/cann-bench/gelu/gelu.py [hunk 14]
```diff
@@ -0,0 +1,515 @@
...
+    else:
+        tiles_per_block = max(2, min(8, (total_tiles + _TOTAL_AIV - 1) // _TOTAL_AIV))
+        while tiles_per_block > 1 and ub_per_tile * tiles_per_block > _UB_BUDGET:
+            tiles_per_block -= 1
+    num_blocks = (total_tiles + tiles_per_block - 1) // tiles_per_block
+    return tiles_per_block, num_blocks
+
+
+@tilelang.jit(out_idx=[-1], pass_configs=pipe_configs)
+def _gelu_tanh_kernel(N, block_N, dtype="float32"):
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:54]: 与 _gelu_tanh_cast_kernel 只差了 cast, 可以使用编译期参数来归一

File: examples/cann-bench/gelu/gelu.py [hunk 15]
```diff
@@ -0,0 +1,515 @@
...
+                T.set_flag("mte3", "mte2", last_stage)
+
+                T.wait_flag("mte3", "mte2", 0)
+                T.wait_flag("mte3", "mte2", 1)
+
+    return main
+
+
+@tilelang.jit(out_idx=[-1], pass_configs=pipe_configs)
+def _gelu_exact_as_kernel(N, block_N, dtype="float32"):
```
Comment (fengz72) [inline examples/cann-bench/gelu/gelu.py:191]: 与 _gelu_exact_as_cast_kernel 只差了 cast, 使用编译期参数归一, 在kernel侧判断是否需要cast

**Review by fengz72 — COMMENTED** [review]
此算子是vector算子, 可以使用pto后端测试性能, 以避免头开销
